### ISSUE TITLE
Processes get more and more when resuing the same pool with timeout

### ISSUE THREAD

**Issue by Maximilianxu** [noxdafox/pebble#72]
Hi
I have a program like the following:
```Python
pool = ProcessPool(max_workers=4)
def do_some_work():
  futs = []
  for _ in range(4):
    fut = pool.schedule(work_func, args=[...], timeout=8)
    futs.append(fut)
  for fut in futs:
    rslt = fut.result()
```
The function do_some_work will be invoked repeatly. I expect that there will be at most 4 processes at the same time. However, the process are getting more and more, and finally consume the resources exhaustly.

Are there any solutions to this?

**Comment by noxdafox** [comment]
Hello,

what is `work_func` doing? Is it spawning processes itself?

**Comment by Maximilianxu** [comment]
@noxdafox 
Hi, `work_func` will compile a program by invoking a C++ function (maybe multiple threads during compiling).

However, if I don't reuse the ProcessPool, and use `with ProcessPool as pool` context in `do_some_work` function instead, there will be at most 4 subprocesses. What I am concerned is that re-allocating a new pool whenever calling the `do_some_work` function will induce much overhead.

**Comment by noxdafox** [comment]
How is the program compilation actually happening?

Are you sure the `work_func` is not spawning a process itself?

Pebble manages the worker processes but it is not responsible of any further resource the provided functions allocate. 
For example, if `work_func` spawns a process itself using `subprocess.Popen`, such process will not be terminated. In case of timeout, the worker process is terminated and the subprocess will remain orphan.
Same applies with files and sockets etc..

**Comment by Maximilianxu** [comment]
I checked the compiling process, and it won't spawn new processes actually. Quite strange, but I re-allocate a new pool in the `work_func` now to work around this. 
Also, I set the multiprocess context as `spawn` by 
```Python
import torch.multiprocessing as multi
multi.set_start_method("spawn", force=True)
``` to block some exceptions during compiling cuda program because setting `fork` will induce a crash of the main process (yet another strange problem, but I didn't have the time yet to check why it is...).
Thanks anyway. If I have more information after check, I will report to you.
